File: frontend/src/lib/schemas/regime-modifier.ts
import { z } from 'zod';

export const geopoliticalStateSchema = z.enum([
  'NONE',
  'RESOLVED',
  'DE_ESCALATING',
  'ACTIVE_RISK',
  'ESCALATING',
]);
export type GeopoliticalState = z.infer<typeof geopoliticalStateSchema>;

// ---------------------------------------------------------------------------
// Top-level response schema
// ---------------------------------------------------------------------------

export const regimeModifierResponseSchema = z.object({
  /** Ticker symbol (upper-case) */
  ticker: z.string(),

  /** Morning briefing geopolitical gate shared across the app */
  geopolitical_state: geopoliticalStateSchema,

  /** Brent crude price in USD per barrel (most recent daily close), or null */
  brent_price: z.number().nullable(),

  /** CBOE VIX index level (most recent daily close), or null */
  vix_value: z.number().nullable(),

  /** Number of consecutive most recent Brent closes below $95 */
  brent_consecutive_below_95_count: z.number().int().min(0),

  /** Original Framework Score before regime adjustment 0–100 */
  base_score: z.number().int().min(0).max(100),

  /** Score after applying the triggered regime rule 0–100 */
  adjusted_score: z.number().int().min(0).max(100),

  /**
   * Which rule fired: 1 = Crisis Halt, 2 = Caution, 3 = Soft Caution, 4 = Clear.
   * null when normal market conditions — no rule triggered.
   */
  rule_triggered: z.union([
    z.literal(1),
    z.literal(2),
    z.literal(3),
    z.literal(4),
  ]).nullable(),

  /** Human-readable name of the triggered rule, e.g. CRISIS | CAUTION | CLEAR | NORMAL. */
  rule: z.string(),

  /** Displayed regime after applying the geopolitical gate. */
  effective_regime: z.string(),

  /** Score delta applied by the triggered rule: -10, -5, -3, +5, or 0. */
  modifier: z.number().int(),

  /** Minimum required cash as a fraction of position value (0.35 = 35%) */
  min_cash_pct: z.number(),

  /** Maximum recommended cash as a fraction of position value (0.40 = 40%) */
  max_cash_pct: z.number(),

  /** Minimum cash in USD — null when ticker is not in the portfolio */
  min_cash_usd: z.number().nullable(),

  /** Maximum cash in USD — null when ticker is not in the portfolio */
  max_cash_usd: z.number().nullable(),

  /** Human-readable cash management instruction from the triggered rule */
  output_text: z.string(),

  /** Explanation of how the regime was determined. */
  determination_text: z.string(),

  // ── Enriched condition display (Section 14 spec v2.1) ─────────────────

  /** Brent price with zone label, e.g. '$97.50 — $95-110 (CAUTION trigger)'. */
  brent_condition: z.string().default(''),

  /** VIX level with zone label, e.g. '17.48 — Below 22 (SOFT CAUTION zone)'. */
  vix_condition: z.string().default(''),

  /** Active geopolitical flag value, e.g. 'ACTIVE_RISK'. */
  geo_condition: z.string().default(''),

  /** Whether this regime uses OR or AND logic. */
  trigger_logic: z.string().default(''),

  /** Explanation of the modifier applied, e.g. 'CAUTION + Escalating geo → −7'. */
  modifier_reason: z.string().default(''),

  /** True only when regime=CAUTION and geo=ESCALATING (−7 modifier). */
  special_case_active: z.boolean().default(false),

  /** Minimum portfolio cash floor fraction per Section 14.1 (0.08 = 8%). */
  cash_floor_pct: z.number().default(0.2),
});

// ---------------------------------------------------------------------------
// Derived type
// ---------------------------------------------------------------------------

export type RegimeModifierResponse = z.infer<typeof regimeModifierResponseSchema>;

File: backend/src/atlas/schemas/regime_modifier.py
"""Pydantic schemas for the Regime Modifier endpoint."""

from pydantic import BaseModel, Field
from typing import Literal


GeopoliticalState = Literal["NONE", "RESOLVED", "DE_ESCALATING", "ACTIVE_RISK", "ESCALATING"]


class RegimeModifierResponse(BaseModel):
    """Response from the regime modifier endpoint."""

    ticker: str
    geopolitical_state: GeopoliticalState

    # ── Market conditions (None when Polygon.io data unavailable) ─────────
    brent_price: float | None = Field(
        description="Brent crude price in USD per barrel (most recent daily close).",
    )
    vix_value: float | None = Field(
        description="CBOE VIX index level (most recent daily close).",
    )
    brent_consecutive_below_95_count: int = Field(
        description="Number of consecutive Brent closes below $95 using the most recent closes.",
    )

    # ── Framework score ───────────────────────────────────────────────────
    base_score: int = Field(
        description="Original framework score before regime adjustment (0-100).",
    )
    adjusted_score: int = Field(
        description="Score after applying the triggered regime rule (0-100).",
    )

    # ── Regime determination ──────────────────────────────────────────────
    rule_triggered: int | None = Field(
        description="Which regime rule fired (1 = Crisis, 2 = Caution, 3 = Clear), "
        "or None when market conditions are normal.",
    )
    rule: str = Field(
        description="Human-readable name of the triggered rule: "
        "CRISIS | CAUTION | CLEAR | NORMAL.",
    )
    effective_regime: str = Field(
        description="Displayed regime after applying the geopolitical gate.",
    )
    modifier: int = Field(
        description="Score delta applied by the triggered rule: "
        "-10 (Crisis), -5 (Caution), +5 (Clear), 0 (Normal).",
    )

    # ── Cash guidance — fractions (0.35 = 35 %) ──────────────────────────
    min_cash_pct: float = Field(
        description="Minimum required cash fraction of the ticker's position value.",
    )
    max_cash_pct: float = Field(
        description="Maximum recommended cash fraction of the ticker's position value.",
    )

    # ── Cash guidance — USD amounts ───────────────────────────────────────
    # None when the ticker is not present in the portfolio DB.
    min_cash_usd: float | None = Field(
        description="Minimum cash in USD (min_cash_pct x position_value). "
        "None when ticker is not in the portfolio.",
    )
    max_cash_usd: float | None = Field(
        description="Maximum cash in USD (max_cash_pct x position_value). "
        "None when ticker is not in the portfolio.",
    )

    # ── Human-readable guidance ───────────────────────────────────────────
    output_text: str = Field(
        description="Regime cash-management instruction (multi-line for crisis rules).",
    )
    determination_text: str = Field(
        description="Human-readable summary of the Brent/VIX calculation and modifier.",
    )

    # ── Enriched condition display (Section 14 spec v2.1) ─────────────────
    brent_condition: str = Field(
        description="Brent price with zone label, e.g. '$97.50 — $95-110 (CAUTION trigger)'.",
        default="",
    )
    vix_condition: str = Field(
        description="VIX level with zone label, e.g. '17.48 — Below 22 (SOFT CAUTION zone)'.",
        default="",
    )
    geo_condition: str = Field(
        description="Active geopolitical flag value, e.g. 'ACTIVE_RISK'.",
        default="",
    )
    trigger_logic: str = Field(
        description="Whether this regime uses OR or AND logic: 'OR — either Brent or VIX triggers'.",
        default="",
    )
    modifier_reason: str = Field(
        description="Explanation of the modifier applied, e.g. 'CAUTION + Escalating geo → −7'.",
        default="",
    )
    special_case_active: bool = Field(
        description="True only when regime=CAUTION and geo=ESCALATING (−7 modifier).",
        default=False,
    )
    cash_floor_pct: float = Field(
        description="Minimum portfolio cash floor fraction per Section 14.1 (0.08 = 8%).",
        default=0.20,
    )
